hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/CustomGasCharging.java

```diff
@@ -43,15 +62,78 @@ public CustomGasCharging(@NonNull final GasCalculator gasCalculator) {
-     * @return the result of the gas charging
-     * @throws HandleException if the gas charging fails fo
+     * @return the amount charged to the relayer, if any
+     * @throws HandleException if the gas charging fails for any reason
      */
-    public GasChargingResult tryToChargeGas(
+    public long chargeForGas(
             @NonNull final HederaEvmAccount sender,
             @Nullable final HederaEvmAccount relayer,
             @NonNull final HederaEvmContext context,
             @NonNull final HederaWorldUpdater worldUpdater,
             @NonNull final HederaEvmTransaction transaction) {
-        throw new AssertionError("Not implemented");
+        if (context.staticCall()) {
+            return 0L;
+        }
+        final var intrinsicGas = gasCalculator.transactionIntrinsicGasCost(Bytes.EMPTY, transaction.isCreate());
+        validateTrue(transaction.gasLimit() >= intrinsicGas, INSUFFICIENT_GAS);
+        if (transaction.isEthereumTransaction()) {
+            return chargeWithRelayer(sender, requireNonNull(relayer), context, worldUpdater, transaction);
+        } else {
+            chargeWithOnlySender(sender, context, worldUpdater, transaction);
+            return 0L;
+        }
+    }
+
+    private void chargeWithOnlySender(
+            @NonNull final HederaEvmAccount sender,
+            @NonNull final HederaEvmContext context,
+            @NonNull final HederaWorldUpdater worldUpdater,
+            @NonNull final HederaEvmTransaction transaction) {
+        final var gasCost = transaction.gasCostGiven(context.gasPrice());
+        final var upfrontCost = transaction.upfrontCostGiven(context.gasPrice());
+        // We validate up-front cost here just for consistency with existing code
+        validateTrue(sender.getBalance().toLong() >= upfrontCost, INSUFFICIENT_PAYER_BALANCE);
+        validateAndCharge(gasCost, sender, worldUpdater);
+    }
+
+    private long chargeWithRelayer(
+            @NonNull final HederaEvmAccount sender,
+            @NonNull final HederaEvmAccount relayer,
+            @NonNull final HederaEvmContext context,
+            @NonNull final HederaWorldUpdater worldUpdater,
+            @NonNull final HederaEvmTransaction transaction) {
+        final var gasCost = transaction.gasCostGiven(context.gasPrice());
+        if (transaction.requiresFullRelayerAllowance()) {
+            validateTrue(transaction.maxGasAllowance() >= gasCost, INSUFFICIENT_TX_FEE);
+            validateAndCharge(gasCost, requireNonNull(relayer), worldUpdater);
+            return gasCost;
+        } else if (transaction.offeredGasPrice() >= context.gasPrice()) {
+            validateAndCharge(gasCost, sender, worldUpdater);
+            return 0L;
+        } else {
+            final var relayerGasCost = gasCost - transaction.offeredGasCost();
+            validateTrue(transaction.maxGasAllowance() >= relayerGasCost, INSUFFICIENT_TX_FEE);
+            validateAndCharge(
```

Comment:
Is it clearer to call the 3 argument version twice?  I can be convinced either way.

Reply:
I just didn't want to have worry about a case where one payer could afford their share but not the other; so added this variant to preserve current behavior.